### PR TITLE
HPCC-15258 Thor stranded support + stranded project

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith jake.smith@lexisnexis.com

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - please review.
NB: I am investigating a failure in teststdlibrary.ecl.
Other than that, all regression tests pass, with forceNumStrands=4

**Comment by richardkchapman** [comment]
There's a #if 1 in thtraceslave.hpp that needs sorting out (can't comment on the file diff because too big )

**Comment by ghalliday** [comment]
Comments will need to be out of line because the diff is too large:

CSlaveActivity:
Minor: Should some of the base classes be contained rather than inherited?  They're not really is-a relationships.
There are a large number of arrays in the class.  It looks like they should be one or two arrays of objects - it would clarify how they were grouped together.  Are they associated with inputs? streams??

CThorStrandProcessor
- count is not a great name - a count of what.  Assuming it is a record count, it is a bad idea for count to be atomic.  It should only be updated from a single thread.

**Comment by ghalliday** [comment]
CThorStarndProcessor::stop()  parent.stop().  Potential for confusion since for some activities parent also implements the rowengine interface - is it stopping the activity or the stream?

**Comment by jakesmith** [comment]
>There are a large number of arrays in the class. It looks like they should be one or two arrays of objects - it would clarify how they were grouped together. Are they associated with inputs? streams??

Agree. There's a comment saying I should change to array of objects, but given the time pressure..

**Comment by jakesmith** [comment]
>count is not a great name - a count of what. Assuming it is a record count, it is a bad idea for count to be atomic. It should only be updated from a single thread.

I'll look / change

**Comment by ghalliday** [comment]
thgraphslave.cpp(317) comment is out date/wrong place
CThorStrandedActivity - strands I was going to rename to processors in roxie.

CSlaveActivity::setInputStream
-             inputStream = tracingStream;
I suspect this should only be done if idx==0?

_input->setOutputStream(inputOutIdx, LINK(inputStream));
   should the second parameter be _inputStream?

**Comment by ghalliday** [comment]
CSlaveActivity::reset()
- this seems to be clearing junction and streams each time.  In roxie they are set up once, and then always reuse the same ones.  That is likely to impose an extra overhead on child queries.

**Comment by ghalliday** [comment]
BaseChooseSetsActivity::stop - can be deleted?

**Comment by ghalliday** [comment]
The asynccall code should be commoned up - a separate PR though.

**Comment by ghalliday** [comment]
CProjectStrandProcessor could switch over to using numProcessedLastGroup in the base class (allows for a slightly more efficient implementation)

**Comment by ghalliday** [comment]
CProjectSlaveActivity - no need to IMPLEMENT_IINTERFACE...

**Comment by ghalliday** [comment]
Finished an initial scan of the code.  Generally it looks clean, and has improved some of the existing code.  A few issues above, some minor, others more significant.

**Comment by jakesmith** [comment]
>CProjectSlaveActivity - no need to IMPLEMENT_IINTERFACE...

there are quite a few others that can be removed. I'll do that.

**Comment by jakesmith** [comment]
>BaseChooseSetsActivity::stop - can be deleted?

yes, will delete

**Comment by jakesmith** [comment]
>thgraphslave.cpp(317) comment is out date/wrong place
>CThorStrandedActivity - strands I was going to rename to processors in roxie.

Which line is that?
line 317 in this PR is a line with just a closing brace on it.

**Comment by jakesmith** [comment]
```
CSlaveActivity::setInputStream

inputStream = tracingStream; I suspect this should only be done if idx==0?
_input->setOutputStream(inputOutIdx, LINK(inputStream));
should the second parameter be _inputStream?
```

a few things were wrong here, have fixed, see next commit.

**Comment by jakesmith** [comment]
>The asynccall code should be commoned up - a separate PR though.

Have made common (to Thor for now) and used std::function and std::bind to generalize the method passing to utility class.

**Comment by jakesmith** [comment]
>>count is not a great name - a count of what. Assuming it is a record count, it is a bad idea for count to be atomic. It should only be updated from a single thread.

>I'll look / change

Not sure what best strategy and in fact realize that although atomics are not used in general for progress counts, they are potentially not thread safe either.. (when progress is serialized).

Unlike roxie, which only merges in the count (from strands) at end (at stop()), Thor's progress is on a timer, we wouldn't want to use a mutex obviously.

I could a separate count atomically, expressly for the purposes of serializing every N records.
IOW the serialized count would lag behind actual count.

@ghalliday - what do you think?

**Comment by jakesmith** [comment]
>CThorStarndProcessor::stop() parent.stop(). Potential for confusion since for some activities parent also implements the rowengine interface - is it stopping the activity or the stream?

I agree, but borrowed code for roxie implementation.
This is the active count stop thing. The strands call parent.stop in stranded activity until count is zero then call activity stop.
The methods and parent type could be more explicit.
I'll look at changing.

**Comment by jakesmith** [comment]
>CProjectStrandProcessor could switch over to using numProcessedLastGroup in the base class (allows for a slightly more efficient implementation)

done.

File: thorlcr/graph/thgraphslave.cpp [hunk 1]
```diff
@@ -152,63 +149,52 @@ void CSlaveActivity::setInput(unsigned index, CActivityBase *inputActivity, unsi
...
     ForEachItemIn(index, inputs)
     {
-		IThorDataLink *_input = inputs.item(index);
-		if (_input)
-		{
-			unsigned inputOutIdx = inputSourceIdxs.item(index);
-			setInputStream(index, _input, inputOutIdx, consumerOrdered);
-		}
+        CThorInput &_input = inputs.item(index);
+		if (_input.itdl)
```
Comment (ghalliday) [inline thorlcr/graph/thgraphslave.cpp]: minor: tabs sneaked into the source..  actually were there already.

**Comment by ghalliday** [comment]
CProjectStrandProcessor::CProjectStrandProcessor
- currently calling parent.queryRowAllocator().  That may lead to extra contention between the strands.  Roxie creates a unique allocator per strand.  We could leave as it is, but we should open a JIRA to discuss exactly what we should do about allocators for strands.  (Some are cached in the helper and so they will never be unique per strand.)

**Comment by ghalliday** [comment]
"CProjectStrandProcessor could switch over to using numProcessedLastGroup in the base class (allows for a slightly more efficient implementation)" -> done
I don't think you have pushed this change.

**Comment by ghalliday** [comment]
whitespace: CThorStrandOptions::CThorStrandOptions before blockSize

**Comment by ghalliday** [comment]
CThorNarySlaveActivity::start()
- is the change to call startAllInputs(), rather than start on each expanded input correct?

Same for the change in stop...

**Comment by jakesmith** [comment]
>I don't think you have pushed this change.

yep, I forgot to push it, have now

**Comment by ghalliday** [comment]
separate PR, but CEdgeProgress should be able to only increment icount, and add icount to count when returning the value.

**Comment by ghalliday** [comment]
CThorInput: could add an initialiser onto sourceIdx

**Comment by ghalliday** [comment]
CThorInput move sourceIdx as the first element to reduce memory consumption by 8 bytes.

**Comment by jakesmith** [comment]
>currently calling parent.queryRowAllocator(). That may lead to extra contention between the strands. Roxie creates a unique allocator per strand. We could leave as it is, but we should open a JIRA to discuss exactly what we should do about allocators for strands. (Some are cached in the helper and so they will never be unique per strand.)

yes, I meant to, I had forgotten, I think worth doing now.

**Comment by ghalliday** [comment]
CThorStrandProcessor: rowsProcessed should be initialised to 0

**Comment by ghalliday** [comment]
@jakesmith Finished looking through again.  Needs rebasing and a few more comments above, but looking close to ready.

**Comment by jakesmith** [comment]
>yes, I meant to, I had forgotten, I think worth doing now.

have pushed that.

**Comment by jakesmith** [comment]
>CThorStrandProcessor: rowsProcessed should be initialised to 0

it already is, in the constructor. Though I do prefer member initialize style now.

**Comment by jakesmith** [comment]
>CThorInput: could add an initialiser onto sourceIdx
>CThorInput move sourceIdx as the first element to reduce memory consumption by 8 bytes

will do.

**Comment by richardkchapman** [comment]
This is going to need rebasing for a clean merge

**Comment by jakesmith** [comment]
>is the change to call startAllInputs(), rather than start on each expanded input correct?

Possibly not, though Roxie does the same.
Thor used to just start the expandedInputs

@richardkchapman  - should Roxie/Thor only be starting and stopping the expanded inputs?

**Comment by richardkchapman** [comment]
I don't understand the question...

**Comment by jakesmith** [comment]
Thor's Nary activity used to start/stop just the expanded inputs.
It now starts/stops all.
Roxie also starts/stop all.

I'd have thought it should have only started/stopped the concrete inputs.
I'm going to change Thor back, but should Roxie change?

**Comment by jakesmith** [comment]
a couple more commits added
@ghalliday - can u review, then I'll squash

**Comment by ghalliday** [comment]
slave.ipp NaryInputActivity::stop()
- you should be calling stop on the expandedStreams rather then the input activities.

**Comment by ghalliday** [comment]
@jakesmith one comment

**Comment by jakesmith** [comment]
>you should be calling stop on the expandedStreams rather then the input activities.

right.

I will change, then squash the commits.

**Pull request merged**